### PR TITLE
Refactoring of erase and write operations in bs_upload

### PR DESCRIPTION
**Pull request by de-nordic**
Two commits: 
 1) refactoring of erase logic;
 2) refactoring of write logic;

~~Depends on https://github.com/mcu-tools/mcuboot/pull/1388~~

### CODE REVIEW

**Review by nvlsianpu — CHANGES_REQUESTED** [review]

File: boot/boot_serial/src/boot_serial.c [hunk 1]
```diff
@@ -265,6 +265,54 @@ bs_list(char *buf, int len)
...
+{
+    struct flash_sector sect;
+    size_t size;
+    int rc;
+
+    if (end >= flash_area_get_size(fap)) {
+        return -EINVAL;
+    }
+
+    if (end < start) {
```
[the remark this answers is not in the document]
Comment (de-nordic) [reply boot/boot_serial/src/boot_serial.c:295]: Ha. Thanks for pointing this out, this allows me to shave some additional "if"s outside.
This cuts down few more bytes from flash

File: boot/boot_serial/src/boot_serial.c [hunk 2]
```diff
@@ -488,29 +488,18 @@ bs_upload(char *buf, int len)
...
 
-        if (img_chunk_len) {
-            memcpy(wbs_aligned, img_chunk, rem_bytes);
-            memset(wbs_aligned + rem_bytes, flash_area_erased_val(fap),
-                   sizeof(wbs_aligned) - rem_bytes);
-            rc = flash_area_write(fap, curr_off, wbs_aligned, flash_area_align(fap));
-        }
-    } else {
-        rc = flash_area_write(fap, curr_off, img_chunk, img_chunk_len);
+        rc = flash_area_write(fap, curr_off, wbs_aligned, flash_area_align(fap));
```
[the remark this answers is not in the document]
Comment (de-nordic) [reply boot/boot_serial/src/boot_serial.c]: Fixed this.

File: boot/boot_serial/src/boot_serial.c [hunk 3]
```diff
@@ -485,33 +485,23 @@ bs_upload(char *buf, int len)
...
-        }
-    } else {
-        rc = flash_area_write(fap, curr_off, img_chunk, img_chunk_len);
+        rc = flash_area_write(fap, curr_off + img_chunk_len - rem_bytes,
+                              wbs_aligned, flash_area_align(fap));
     }
 
     if (rc == 0) {
-        curr_off += img_chunk_len;
+        curr_off += img_chunk_len + rem_bytes;
```
Comment (nvlsianpu) [inline boot/boot_serial/src/boot_serial.c:504]: `curr_off += img_chunk_len;`
Above code always writing img_chunk_len bytes.

**Review by nvlsianpu — CHANGES_REQUESTED** [review]

File: boot/boot_serial/src/boot_serial.c (hunk 3 above)
Comment (nvlsianpu) [reply boot/boot_serial/src/boot_serial.c:506]: @de-nordic you missed that. This line patch should be dropped.

Comment (de-nordic) [reply boot/boot_serial/src/boot_serial.c:506]: I have changed logic in 481-485. Now the `rem_bytes` is always subtracted from the `img_chunk_len` so after line 482 `img_chunk_len` holds aligned number of bytes. The `rem_bytes` gets reset to 0 for all writes except the final one.
Therefore the addition done here will only add `rem_bytes` for final write, in all other cases only the aligned `img_chunk_len` is added.

Comment (nvlsianpu) [reply boot/boot_serial/src/boot_serial.c:506]: looks OK now

File: boot/boot_serial/src/boot_serial.c [hunk 4]
```diff
@@ -348,111 +401,118 @@ bs_upload(char *buf, int len)
...
+        rc = flash_area_erase(fap, 0, area_size);
         if (rc) {
             goto out_invalid_data;
         }
 #endif
-        img_size = data_len;
-    }
-    if (off != curr_off) {
+
+        not_yet_erased = 0;
```
Comment (nvlsianpu) [inline boot/boot_serial/src/boot_serial.c:449]: need to be moved into MCUBOOT_ERASE_PROGRESSIVELY ifdef-ry

**Comment by nvlsianpu** [comment]
@de-nordic You can drop the 1st commit now : #1388 was merged.

**Review by nvlsianpu — APPROVED** [review]

**Pull request merged**